File: codes/go/chapter_sorting/merge_sort/merge_sort.go
package merge_sort
// File: merge_sort.go
// Created Time: 2022-12-13
// Author: msk397 (machangxinq@gmail.com)

/*
	合并左子数组和右子数组
	左子数组区间 [left, mid]
	右子数组区间 [mid + 1, right]
*/
func merge(nums []int, left, mid, right int) {
	// 初始化辅助数组 借助 copy模块
	tmp := make([]int, right-left+1)
	for i := left; i <= right; i++ {
		tmp[i-left] = nums[i]
	}
	// 左子数组的起始索引和结束索引
	left_start, left_end := left-left, mid-left
	// 右子数组的起始索引和结束索引
	right_start, right_end := mid+1-left, right-left
	// i, j 分别指向左子数组、右子数组的首元素
	i, j := left_start, right_start
	// 通过覆盖原数组 nums 来合并左子数组和右子数组
	for k := left; k <= right; k++ {
		// 若 “左子数组已全部合并完”，则选取右子数组元素，并且 j++
		if i > left_end {
			nums[k] = tmp[j]
			j++
		// 否则，若 “右子数组已全部合并完” 或 “左子数组元素 < 右子数组元素”，则选取左子数组元素，并且 i++
		} else if j > right_end || tmp[i] <= tmp[j] {
			nums[k] = tmp[i]
			i++
		// 否则，若 “左子数组元素 > 右子数组元素”，则选取右子数组元素，并且 j++
		} else {
			nums[k] = tmp[j]
			j++
		}
	}
}


func mergeSort(nums []int, left, right int) {
	// 终止条件
	if left >= right {
		return
	}
	// 划分阶段
	mid := (left + right) / 2
	mergeSort(nums, left, mid)
	mergeSort(nums, mid+1, right)
	// 合并阶段
	merge(nums, left, mid, right)
}

File: ru/codes/java/chapter_sorting/merge_sort.java
/**
 * File: merge_sort.java
 * Created Time: 2022-11-25
 * Author: krahets (krahets@163.com)
 */

package chapter_sorting;

import java.util.*;

public class merge_sort {
    /* Объединить левый и правый подмассивы */
    static void merge(int[] nums, int left, int mid, int right) {
        // Диапазон левого подмассива: [left, mid], диапазон правого подмассива: [mid+1, right]
        // Создать временный массив tmp для хранения результата слияния
        int[] tmp = new int[right - left + 1];
        // Инициализировать начальные индексы левого и правого подмассивов
        int i = left, j = mid + 1, k = 0;
        // Пока в левом и правом подмассивах еще есть элементы, сравнивать их и копировать меньший во временный массив
        while (i <= mid && j <= right) {
            if (nums[i] <= nums[j])
                tmp[k++] = nums[i++];
            else
                tmp[k++] = nums[j++];
        }
        // Скопировать оставшиеся элементы левого и правого подмассивов во временный массив
        while (i <= mid) {
            tmp[k++] = nums[i++];
        }
        while (j <= right) {
            tmp[k++] = nums[j++];
        }
        // Скопировать элементы временного массива tmp обратно в соответствующий диапазон исходного массива nums
        for (k = 0; k < tmp.length; k++) {
            nums[left + k] = tmp[k];
        }
    }

    /* Сортировка слиянием */
    static void mergeSort(int[] nums, int left, int right) {
        // Условие завершения
        if (left >= right)
            return; // Завершить рекурсию, когда длина подмассива равна 1
        // Этап разбиения
        int mid = left + (right - left) / 2; // Вычислить середину
        mergeSort(nums, left, mid); // Рекурсивно обработать левый подмассив
        mergeSort(nums, mid + 1, right); // Рекурсивно обработать правый подмассив
        // Этап слияния
        merge(nums, left, mid, right);
    }

    public static void main(String[] args) {
        /* Сортировка слиянием */
        int[] nums = { 7, 3, 2, 6, 0, 1, 5, 4 };
        mergeSort(nums, 0, nums.length - 1);
        System.out.println("После сортировки слиянием nums = " + Arrays.toString(nums));
    }
}
